From fb05072dfc666114c45725d6e9429dd3dee4a19b Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:21 -0400
Subject: [fs] nfsd: move unhash_client_locked call into mark_client_expired_locked

Message-id: <1410898849-22126-138-git-send-email-bfields@redhat.com>
Patchwork-id: 95735
O-Subject: [Patch RHEL7 137/165] nfsd: move unhash_client_locked call into mark_client_expired_locked
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

All the callers except for the fault injection code call it directly
afterward, and in the fault injection case it won't hurt to do so
anyway.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 97403d95e1a7f5b257e90aad1f3284953bc72671
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 8184800..56999cb 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -129,14 +129,6 @@ static bool is_client_expired(struct nfs4_client *clp)
  return clp->cl_time == 0;
 }
 
-static __be32 mark_client_expired_locked(struct nfs4_client *clp)
-{
- if (atomic_read(&clp->cl_refcount))
-  return nfserr_jukebox;
- clp->cl_time = 0;
- return nfs_ok;
-}
-
 static __be32 get_client_locked(struct nfs4_client *clp)
 {
  struct nfsd_net *nn = net_generic(clp->net, nfsd_net_id);
@@ -1628,6 +1620,14 @@ unhash_client(struct nfs4_client *clp)
  spin_unlock(&nn->client_lock);
 }
 
+static __be32 mark_client_expired_locked(struct nfs4_client *clp)
+{
+ if (atomic_read(&clp->cl_refcount))
+  return nfserr_jukebox;
+ unhash_client_locked(clp);
+ return nfs_ok;
+}
+
 static void
 __destroy_client(struct nfs4_client *clp)
 {
@@ -2498,7 +2498,6 @@ nfsd4_create_session(struct svc_rqst *rqstp,
    status = mark_client_expired_locked(old);
    if (status)
     goto out_free_conn;
-   unhash_client_locked(old);
   }
   move_to_confirmed(unconf);
   conf = unconf;
@@ -3044,7 +3043,6 @@ nfsd4_setclientid_confirm(struct svc_rqst *rqstp,
    status = mark_client_expired_locked(old);
    if (status)
     goto out;
-   unhash_client_locked(old);
   }
   move_to_confirmed(unconf);
   conf = unconf;
@@ -4183,7 +4181,6 @@ nfs4_laundromat(struct nfsd_net *nn)
     clp->cl_clientid.cl_id);
    continue;
   }
-  unhash_client_locked(clp);
   list_add(&clp->cl_lru, &reaplist);
  }
  spin_unlock(&nn->client_lock);
-- 
1.7.1